worker/uniter/uniter.go

```diff
@@ -608,6 +617,21 @@ func (u *Uniter) charmState() (bool, *corecharm.URL, int, error) {
 		return canApplyCharmProfile, charmURL, charmModifiedVersion, errors.Trace(err)
 	}
 
+	charmDirExists, err := u.charmDirExists()
```

Comment:
Could we instead just use charm.ReadCharmDir and if it fails then run the upgrade?

Reply:
+1